benchmark/src/main/scala/fs2/benchmark/FlowInteropBenchmark.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+package benchmark
+
+import cats.effect.IO
+import cats.effect.unsafe.implicits.global
+
+import org.openjdk.jmh.annotations.{
+  Benchmark,
+  BenchmarkMode,
+  Mode,
+  OutputTimeUnit,
+  Param,
+  Scope,
+  State
+}
+
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.Flow.{Publisher, Subscriber, Subscription}
+
+import scala.concurrent.Future
+
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class FlowInteropBenchmark {
+  @Param(Array("1024", "5120", "10240"))
+  var totalElements: Int = _
+
+  @Param(Array("1000"))
+  var iterations: Int = _
+
+  @Benchmark
+  def fastPublisher(): Unit = {
+    def publisher =
+      new Publisher[Unit] {
+        override final def subscribe(subscriber: Subscriber[? >: Unit]): Unit =
+          subscriber.onSubscribe(
+            new Subscription {
+              @volatile var i: Int = 0
+              @volatile var canceled: Boolean = false
+
+              override final def request(n: Long): Unit = {
+                Future {
+                  var j = 0
+                  while ((j < n) && (i < totalElements) && !canceled) {
+                    subscriber.onNext(())
+                    i += 1
+                    j += 1
+                  }
+
+                  if (i == totalElements || canceled) {
+                    subscriber.onComplete()
+                  }
```

Comment:
We are doing multiple volatile reads/writes of `i` in this loop. I think we can just copy it into a local variable and then copy it back out.

```suggestion
                  var i = this.i
                  while ((j < n) && ((i < totalElements) && !canceled) {
                    subscriber.onNext(())
                    i += 1
                    j += 1
                  }

                  if (i == totalElements || canceled) {
                    subscriber.onComplete()
                  }
                  
                  this.i = i
```

Reply:
Maybe make this lore obvious and call the local variable "locali" and not rely on shadowing

Reply:
After more discussion in **Discord** we decided to go with a sequential publisher for simplicity.